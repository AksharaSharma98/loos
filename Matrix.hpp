/*
  Matrix.hpp

  A simple matrix wrapper class.  This is not meant to be a mathematical matrix class!
*/

/*
  This file is part of LOOS.

  LOOS (Lightweight Object-Oriented Structure library)
  Copyright (c) 2008, Tod D. Romo, Alan Grossfield
  Department of Biochemistry and Biophysics
  School of Medicine & Dentistry, University of Rochester

  This package (LOOS) is free software: you can redistribute it and/or modify
  it under the terms of the GNU General Public License as published by
  the Free Software Foundation under version 3 of the License.

  This package is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU General Public License for more details.

  You should have received a copy of the GNU General Public License
  along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/



#if !defined(LOOS_MATRIX_HPP)
#define LOOS_MATRIX_HPP

#include <boost/shared_array.hpp>

namespace loos {

  //! Standardize passing of 2-d matrix coords  (row, col)
  struct Duple {
    Duple(const int a, const int b) : j(a), i(b) { }
    Duple() : j(0), i(0) { }

    friend ostream& operator<<(ostream& os, const Duple& d) {
      os << "Duple(" << d.i << "," << d.j << ")";
      return(os);
    }

    int j, i;
  };



  // These are the policy classes for the Matrix class.  They define
  // how the data is actually stored internally, i.e. lower
  // triangular, column major, or row major order...
  // Operator() is where all the magic occurs...  It takes a 2-d
  // matrix coordinate and converts it into a linear index into the
  // array of data...

  //! Class for storing a symmetric triangular matrix
  class Triangular {
  public:
    Triangular(const int y, const int x) : m(y), n(x), s( (y*(y+1))/2 ) {
      assert(y == x && "Cannot have a non-square triangular matrix...  (you know what I mean!)");
    }

    void set(const int y, const int x) {
      assert(y == x && "Cannot have a non-square triangular matrix...  (you know what I mean!)");
      m=y; n=x;
      s = (y*(y+1))/2;
    }

    unsigned long size(void) const { return(s); }

    long operator()(const int y, const int x) const {
      int b = y;
      int a = x;

      if (x > y) {
	b = x;
	a = y;
      }
      
      return( (b * (b + 1)) / 2 + a );
    }

    
  private:
    int m, n;
    long s;
  };

  //! Class for storing a matrix in column-major order
  class ColMajor {
  public:
    ColMajor(const int y, const int x) : m(y), n(x), s(y*x) { }
    void set(const int y, const int x) { m=y; n=x; s = y*x; }
    long size(void) const { return(s); }

    long operator()(const int y, const int x) const { return(x*m + y); }

  private:
    int m, n;
    long s;
  };


  //! Class for storing a matrix in row-major order...
  class RowMajor {
  public:
    RowMajor(const int y, const int x) : m(y), n(x), s(y*x) { }
    void set(const int y, const int x) { m=y; n=x; s=y*x; }
    long size(void) const { return(s); }

    long operator()(const int y, const int x) const { return(y*n + x); }

  private:
    int m, n;
    long s;
  };

  // Reinterpretation functions...
  template<typename T, class C> class Matrix;
  template<typename T> Matrix<T, RowMajor> reinterpretOrder(const Matrix<T, ColMajor>&);
  template<typename T> Matrix<T, ColMajor> reinterpretOrder(const Matrix<T, RowMajor>&);

  // Default order is column major...


  //! Templatized class for intepreting memory as a matrix stored in
  //! different formats...
  /** This class is a wrapper around a block of data allowing the user
   * to access it as a 2D matrix.  The idea is that the data can
   * internally be stored in a row-major, col-major, or triangular
   * format but the user-interface is always the same.  In addition,
   * some capacity to reinterpret the data as col or row major is
   * provided...
   *
   * Internally, the data is managed by a boost shared pointer
   * (boost::shared_array).  Range checking is provided to all
   * accesses.
   *
   * Newly allocate matrices have each element initialized to 0.
   *
   * Note: metadata is currently unused...
   */

  template<typename T, class Policy = ColMajor>
  class Matrix {
  public:

    //! Unitialized matrix
    Matrix() : m(0), n(0), mi(0), pol(0,0), dptr(0), meta("") { }

    //! Wrap an existing block of data with a Matrix.
    Matrix(T* p, const int b, const int a) : m(b), n(a), mi(a*b), pol(b, a), meta("") { dptr = boost::shared_array<T>(p); }

    //! Wrap an already shared block of data...
    Matrix(boost::shared_array<T>& p, const int b, const int a) : m(b), n(a), mi(a*b), pol(b, a), dptr(p), meta("") { }


    //! Create a new block of data for the requested Matrix
    Matrix(const int b, const int a) : m(b), n(a), mi(a*b), pol(b, a), meta("") { allocate(); }

    //! Deep copy of a matrix...
    Matrix<T, Policy> copy(void) const {
      Matrix<T, Policy> result(m, n);
      result.meta = meta;
      long size = pol.size();
      T* p = result.dptr.get();
      T* q = dptr.get();

      for (long i=0; i<size; i++)
	p[i] = q[i];
      
      return(result);
    }

    int rows(void) const { return(m); }
    int cols(void) const { return(n); }

    //! Return a pointer to the internal data...
    T* get(void) { return(dptr.get()); }

    //! Treat the matrix as a 1D array
    T& operator[](const long i) {
      assert(i < mi && "Index out of range in Matrix::operator[]");
      return(dptr[i]);
    }

    //! Return the appropriate element (y-rows, x-cols)
    T& operator()(const int y, const int x) {
      long i = pol(y,x);
      assert(i < mi && "Index out of range in Matrix::operator()");
      return(dptr[i]);
    }

    const T& operator[](const long i) const {
      assert(i < mi && "Index out of range in Matrix::operator[]");
      return(dptr[i]);
    }

    const T& operator()(const int y, const int x) const {
      long i = pol(y,x);
      assert(i < mi && "Index out of range in Matrix::operator()");
      return(dptr[i]);
    }

    void metaData(const string& s) { meta = s; }
    string metaData(void) const { return(meta); }

    //! Deallocate data...
    void free(void) { m = n = 0; dptr.reset(); }

    //! Convert a Col-major to Row-major format
    friend Matrix<T, RowMajor> reinterpretOrder<>(const Matrix<T,ColMajor>&);

    //! Convert a Row-major to Col-major format
    friend Matrix<T, ColMajor> reinterpretOrder<>(const Matrix<T,RowMajor>&);

  private:

    void allocate(void) {
      long size = pol.size();
      dptr.reset();
      dptr = boost::shared_array<T>(new T[size]);
      for (long i=0; i<size; i++)
	dptr[i] = 0;
    }

  private:
    int m, n, mi;
    Policy pol;
    boost::shared_array<T> dptr;
    string meta;
  
  };

  template<typename T>
  Matrix<T, RowMajor> reinterpretOrder(const Matrix<T, ColMajor>& A) {
    Matrix<T, RowMajor> result(A.m, A.n);
    result.dptr = A.dptr;
  
    return(result);
  };

  template<typename T>
  Matrix<T, ColMajor> reinterpretOrder(const Matrix<T, RowMajor>& A) {
    Matrix<T, ColMajor> result(A.m, A.n);
    result.dptr = A.dptr;
  
    return(result);
  };

}

#endif
