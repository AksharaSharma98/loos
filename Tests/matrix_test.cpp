#include <iostream>
#include <string>
#include <iterator>
#include <stdexcept>

#include <boost/format.hpp>


#include "Matrix.hpp"

using namespace std;
using namespace lab;

// Quick formatted output of a matrix...
template<class T>
void show(const T& M, const string& s, const string& fmt) {
  cout << s << endl;
  int m = M.rows();
  int n = M.cols();

  for (int j=0; j<m; ++j) {
    for (int i=0; i<n; ++i)
      cout << boost::format(fmt) % M(j, i);
    cout << endl;
  }
}

// Formatted output of the matrix as a linear array...
template<class T>
void showLinear(const T& M, const string& s, const string& fmt) {
  cout << s << endl;
  long n = M.size();
  for (long i=0; i<n; ++i) {
    cout << boost::format(fmt) % M[i];
    cout << endl;
  }
}



int main(int argc, char *argv[]) {

  // How to write/read a matrix...
  if (argc == 2) {
    string name(argv[1]);
    name += ".asc";
    Matrix<float> M(4,4);
    
    int k=0;
    for (int j=0; j<4; ++j)
      for (int i=0; i<4; ++i)
        M(j,i) = k++;
    
    show(M, "M", "%8.2f");
    
    writeAsciiMatrix(name, M, "Testing");
    
    // First form for reading...
    Matrix<float, ColMajor, SharedArray> A = readAsciiMatrix<float, ColMajor, SharedArray>(name);
    cout << boost::format("Read in a %d x %d matrix.\n") % A.rows() % A.cols();
    show(A, "A", "%8.2f");

    // Second form for reading...
    // Relies on default polices...
    Matrix<float> B;
    readAsciiMatrix(name, B);
    cout << boost::format("(2nd form) read in a %d x %d matrix.\n") % B.rows() % B.cols();
    show(B, "B", "%8.2f");
  }

  // Create a row-major matrix with floating point elements...
  Matrix<float, RowMajor> M(4,4);
  int k=0;
  for (int j=0; j<4; ++j)
    for (int i=0; i<4; ++i, ++k)
      M(j,i) = k;

  showLinear(M, "M (row-major)", "%8.2f");

  // You can also output a matrix in Matlab/Octave m-script syntax
  // using the << operator...

  cout << "M = " << M << endl;


  // Create a col-major matrix with integer elements
  Matrix<int, ColMajor> N(4, 4);
  for (int j=0,k=0; j<4; ++j)
    for (int i=0; i<4; ++i, ++k)
      N(j, i) = k;
  show(N, "N (col-major)", "%8.2f");
  showLinear(N, "N (col-major)", "%8.2f");

   // Reinterpret as row-major
  Matrix<float, ColMajor> MM = reinterpretOrder(M);
  show(MM, "N as row-major", "%8.2f");
  showLinear(MM, "MM (linealy)", "%8.2f");

  // Create a triangular matrix...
  Matrix<int, Triangular> T(4,4);
  for (int j=0,k=0; j<4; ++j)
    for (int i=0; i<=j; ++i, ++k)
      T(j, i) = k;
  show(T, "Triangular", "%4d");

  // Accessing the low-level data
  int *ptr = T.get();
  cout << "Low-level access to T\n";
  long s = T.size();
  for (long i=0; i<s; ++i)
    cout << "\t" << ptr[i] << endl;

  // Test of index range-checking...
  bool exception_caught = false;
  try {
    int v = T(4,4);
  }
  catch(out_of_range& e) {
    exception_caught = true;
  }

  if (!exception_caught)
    cout << "***WARNING***\nWe didn't catch an expected out_of_range exception.\n";

  cout << "* iterator test *\n";
  cout << "T = ";
  copy(T.begin(), T.end(), ostream_iterator<int>(cout, ","));
  cout << endl;

  if (argc == 2) {
    cout << "* Writing Triangular Matrix *\n";
    string name(argv[1]);
    name += ".tri";
    writeAsciiMatrix(name, T, "Testing");
    Matrix<int, Triangular> TT;
    readAsciiMatrix(name, TT);
    cout << boost::format("Read in a %d x %d triangular matrix.\n") % TT.rows() % TT.cols();
    show(TT, "T (from file)", "%8.2f");
  }

  cout << "* Copy test *\n";
  M(1,1) = 3.141;
  Matrix<float,RowMajor> MC = M.copy();
  MC(1,1) = 2.718;
  show(M, "Original (1,1)=pi", "%8.2f");
  show(MC, "Copy (1,1)=e", "%8.2f");

  // Sparse test...
  Matrix<float, RowMajor, SparseArray> S(4,4);
  S(1,1) = 1;
  S(2,2) = 3;
  S(3,3) = 5;
  S(1,3) = 7;
  cout << "* Sparse test *\n";
  cout << "actualSize = " << S.actualSize() << endl;
  show(S, "Sparse", "%8.2f");
  cout << "actualSize = " << S.actualSize() << endl;

  // This tests writing/reading of sparse matrices...
  if (argc == 2) {
    string name(argv[1]);
    name += ".spm";

    cout << "* Sparse IO test *\n";
    writeAsciiMatrix(name, S, "Testing");
    Matrix<float, RowMajor, SparseArray> SS;
    readAsciiMatrix(name, SS);

    cout << boost::format("Read in a %d x %d sparse matrix.\n") % SS.rows() % SS.cols();
    cout << "actualSize = " << SS.actualSize() << endl;
    show(SS, "Sparse (read in from file)", "%8.2f");
  }

  // Test of copyMatrix...
  Matrix<float, RowMajor, SparseArray> SC;
  copyMatrix(SC, S);
  cout << "actualSize of copy = " << SC.actualSize() << endl;
  cout << "Density is " << SC.density() << endl;
  
}
