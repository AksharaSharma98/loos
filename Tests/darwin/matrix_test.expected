M
    0.00    1.00    2.00    3.00
    4.00    5.00    6.00    7.00
    8.00    9.00   10.00   11.00
   12.00   13.00   14.00   15.00
Read in a 4 x 4 matrix.
A
    0.00    1.00    2.00    3.00
    4.00    5.00    6.00    7.00
    8.00    9.00   10.00   11.00
   12.00   13.00   14.00   15.00
(2nd form) read in a 4 x 4 matrix.
B
    0.00    1.00    2.00    3.00
    4.00    5.00    6.00    7.00
    8.00    9.00   10.00   11.00
   12.00   13.00   14.00   15.00
M (row-major)
    0.00
    1.00
    2.00
    3.00
    4.00
    5.00
    6.00
    7.00
    8.00
    9.00
   10.00
   11.00
   12.00
   13.00
   14.00
   15.00
M = # 4 4 (0)
0 1 2 3 
4 5 6 7 
8 9 10 11 
12 13 14 15 

N (col-major)
       0       1       2       3
       4       5       6       7
       8       9      10      11
      12      13      14      15
N (col-major)
       0
       4
       8
      12
       1
       5
       9
      13
       2
       6
      10
      14
       3
       7
      11
      15
N as row-major
    0.00    4.00    8.00   12.00
    1.00    5.00    9.00   13.00
    2.00    6.00   10.00   14.00
    3.00    7.00   11.00   15.00
MM (linealy)
    0.00
    1.00
    2.00
    3.00
    4.00
    5.00
    6.00
    7.00
    8.00
    9.00
   10.00
   11.00
   12.00
   13.00
   14.00
   15.00
Triangular
   0   1   3   6
   1   2   4   7
   3   4   5   8
   6   7   8   9
Low-level access to T
	0
	1
	2
	3
	4
	5
	6
	7
	8
	9
* iterator test *
T = 0,1,2,3,4,5,6,7,8,9,
* Writing Triangular Matrix *
Read in a 4 x 4 triangular matrix.
T (from file)
       0       1       3       6
       1       2       4       7
       3       4       5       8
       6       7       8       9
* Copy test *
Original (1,1)=pi
    0.00    1.00    2.00    3.00
    4.00    3.14    6.00    7.00
    8.00    9.00   10.00   11.00
   12.00   13.00   14.00   15.00
Copy (1,1)=e
    0.00    1.00    2.00    3.00
    4.00    2.72    6.00    7.00
    8.00    9.00   10.00   11.00
   12.00   13.00   14.00   15.00
* Sparse test *
actualSize = 4
Sparse
    0.00    0.00    0.00    0.00
    0.00    1.00    0.00    7.00
    0.00    0.00    3.00    0.00
    0.00    0.00    0.00    5.00
actualSize = 4
* Sparse IO test *
Read in a 4 x 4 sparse matrix.
actualSize = 4
Sparse (read in from file)
    0.00    0.00    0.00    0.00
    0.00    1.00    0.00    7.00
    0.00    0.00    3.00    0.00
    0.00    0.00    0.00    5.00
actualSize of copy = 4
Density is 0.25
